pkgs/development/python-modules/generic/wrap.sh

```diff
@@ -1,51 +1,82 @@
+# Wrapper around wrapPythonProgramsIn, below. The $pythonPath
+# variable is passed in from the buildPythonPackage function.
 wrapPythonPrograms() {
     wrapPythonProgramsIn $out "$out $pythonPath"
 }
 
+# Transforms any binaries generated by the setup.py script, replacing them
+# with an executable shell script which will set some environment variables
+# and then call into the original binary (which has been given a .wrapped
+# prefix).
 wrapPythonProgramsIn() {
     local dir="$1"
     local pythonPath="$2"
     local python="@executable@"
-    local i
+    local path
+    local f
 
+    # Create an empty table of python paths (see doc on _addToPythonPath
+    # for how this is used). Build up the program_PATH and program_PYTHONPATH
+    # variables.
     declare -A pythonPathsSeen=()
     program_PYTHONPATH=
     program_PATH=
-    for i in $pythonPath; do
-        _addToPythonPath $i
+    for path in $pythonPath; do
+        _addToPythonPath $path
     done
 
-    for i in $(find "$dir" -type f -perm +0100); do
+    # Find all regular files in the output directory that are executable.
+    for f in $(find "$dir" -type f -perm +0100); do
         # Rewrite "#! .../env python" to "#! /nix/store/.../python".
-        if head -n1 "$i" | grep -q '#!.*/env.*\(python\|pypy\)'; then
-            sed -i "$i" -e "1 s^.*/env[ ]*\(python\|pypy\)^#! $python^"
+        if head -n1 "$f" | grep -q '#!.*/env.*\(python\|pypy\)'; then
+            sed -i "$f" -e "1 s^.*/env[ ]*\(python\|pypy\)^#! $python^"
         fi
-        
+
         # catch /python and /.python-wrapped
-        if head -n1 "$i" | grep -q '/\.\?\(python\|pypy\)'; then
+        if head -n1 "$f" | grep -q '/\.\?\(python\|pypy\)'; then
             # dont wrap EGG-INFO scripts since they are called from python
-            if echo "$i" | grep -v EGG-INFO/scripts; then
-                echo "wrapping \`$i'..."
-                sed -i "$i" -re '@magicalSedExpression@'
-                wrapProgram "$i" \
-                    --prefix PYTHONPATH ":" $program_PYTHONPATH \
-                    --prefix PATH ":" $program_PATH
+            if echo "$f" | grep -qv EGG-INFO/scripts; then
+                echo "wrapping \`$f'..."
+                sed -i "$f" -re '@magicalSedExpression@'
+                # wrapProgram creates the executable shell script described
+                # above. The script will set PYTHONPATH and PATH variables.!
+                # (see pkgs/build-support/setup-hooks/make-wrapper.sh)
+                local wrap_args="$f \
+                                 --prefix PYTHONPATH ':' $program_PYTHONPATH \
+                                 --prefix PATH ':' $program_PATH"
+
+                # Add any additional environment variables to propagate.
+                for env_var in $setEnvVars; do
```

Comment:
`setEnvVars` is an argument to `buildPythonPackage` and is passed through to `mkDerivation`. Since it's a list, it will end up as a `\n`-separated string environment variable (is what I'm given to understand, seeing as it's the same with the `pythonPath` variable elsewhere in this script).

Yes, it does work :smile: